### PR TITLE
Revise item usage and consumption

### PR DESCRIPTION
**Pull request by krbz999**
The function `Item5e#_getUsageConfig` returns an object of what an item is able to do:
- Create a template.
- Consume a spell slot.
- Consume limited uses/recharge.
- Consume a resource.

The function `Item5e#_getUsageConfigValues` returns an object that depends on `_getUsageConfig` to determine what an item, by default, will do. This is the object that can be modified directly in `Item5e#use`. This also includes `slotLevel`, which is now always a string.

Whether an item then needs configuration is determined by if any of the default values are `true` (they are by default `true` if the item is able to fulfill their function).

If none are true, a prompt will not be forced. Any boxes that are possible but overridden to be `false` by default will still appear in the `AbilityUseDialog`, but unchecked; if the prompt is skipped (`options.configureDialog` being `false`) the initial values are used (possibly partially overridden).

Also added two properties in `item.system.uses` and `item.system.target` for #2463 which function with `getUsageConfigValues`.

### CODE REVIEW

File: module/documents/item.mjs [hunk 1]
```diff
@@ -903,6 +904,46 @@ export default class Item5e extends Item {
...
+      consumeUses: active && this.hasLimitedUses,
+      consumeResource: active && !!consume.type && !!consume.target && (!this.hasAttack || (consume.type !== "ammo")),
+      createTemplate: active && game.user.can("TEMPLATE_CREATE") && this.hasAreaTarget
+    };
+
+    // Do not suggest consuming your own uses if also consuming them through resources.
+    if ( config.consumeResource && (consume.target === this.id) ) config.consumeUses = false;
+
+    return config;
+  }
```
Comment (Fyorl) [inline module/documents/item.mjs:926]: It might not be feasible for various reasons that I hadn't considered, but, as discussed on discord, can we consider a design where we do not need both `_getUsageConfig` and `_getUsageConfigValues` but instead roll them into one? Either the resulting config simply omits an option if it does not apply to the item, or sets it to some special value (like `null`).

I would think then the logic for the `AbilityUseDialog` becomes quite simple: If there are no non-`null` options in the config (or no options at all), the dialog can be skipped. If a value is `true`, it is present but checked by default, and if it is `false`, it is present but unchecked by default.

File: module/data/item/templates/activated-effect.mjs [hunk 2]
```diff
@@ -86,7 +87,8 @@ export default class ActivatedEffectTemplate extends foundry.abstract.DataModel
         per: new foundry.data.fields.StringField({
           required: true, nullable: true, blank: false, initial: null, label: "DND5E.LimitedUsesPer"
         }),
-        recovery: new FormulaField({required: true, label: "DND5E.RecoveryFormula"})
+        recovery: new FormulaField({required: true, label: "DND5E.RecoveryFormula"}),
+        prompt: new foundry.data.fields.BooleanField({initial: true, label: "DND5E.LimitedUsesPrompt"})
```
Comment (Fyorl) [inline module/data/item/templates/activated-effect.mjs:91]: It would be ideal if we could keep this PR focused purely on the refactor, and separate out this feature into its own PR once this one is merged, so that it can be discussed separately.

File: module/documents/item.mjs [hunk 3]
```diff
@@ -912,73 +953,88 @@ export default class Item5e extends Item {
    * @returns {object|boolean}             A set of data changes to apply when the item is used, or false.
    * @protected
    */
-  _getUsageUpdates({
-    consumeQuantity, consumeRecharge, consumeResource, consumeSpellSlot,
-    consumeSpellLevel, consumeUsage}) {
+  _getUsageUpdates(config) {
     const actorUpdates = {};
-    const itemUpdates = {};
+    let itemUpdates = {};
```
Comment (Fyorl) [inline module/documents/item.mjs:958]: This only needs to be `let` if the variable is re-assigned somewhere, but I can't see that happening in this method, unless I've missed it?

File: module/documents/item.mjs [hunk 4]
```diff
@@ -912,73 +953,88 @@ export default class Item5e extends Item {
...
-      }
+    // Consume uses (or quantity).
+    else if ( !!uses.max && !!uses.per && (uses.value > 0) ) {
+      const remaining = Math.max(uses.value - 1, 0);
 
-      // If the item was not used, return a warning
-      if ( !used ) {
-        ui.notifications.warn(game.i18n.format("DND5E.ItemNoUses", {name: this.name}));
-        return false;
+      if ( remaining > 0 || ( !remaining && !uses.autoDestroy) ) {
```
Comment (Fyorl) [inline module/documents/item.mjs:1020]:
```suggestion
      if ( remaining > 0 || (!remaining && !uses.autoDestroy) ) {
```

**Review by Fyorl — CHANGES_REQUESTED** [review]
Good changes here, thank you for taking on this refactor. I think you've managed to streamline the logic and make the pipeline a lot easier to follow (at least for me). I have some suggestions for your consideration that we can discuss.

File: module/documents/item.mjs [hunk 5]
```diff
@@ -816,23 +813,26 @@ export default class Item5e extends Item {
...
-      const upcastLevel = config.consumeSpellLevel === "pact" ? as.spells.pact.level
-        : parseInt(config.consumeSpellLevel);
-      if ( upcastLevel && (upcastLevel !== is.level) ) {
-        item = item.clone({"system.level": upcastLevel}, {keepId: true});
+    // Handle upcasting
+    if ( item.type === "spell" ) {
+      let level = null;
+      if ( config.slotLevel ) {
+        // A spell slot was consumed.
+        level = config.slotLevel === "pact" ? as.spells.pact.level : parseInt(config.slotLevel.at(-1));
```
[the remark this answers is not in the document]
Comment (krbz999) [reply module/documents/item.mjs:827]: I'd be happier with a `config.slotLevel.replace("spell", "")` in that case and gain some uniformity for the inputs, all being strings.

File: module/documents/item.mjs [hunk 6]
```diff
@@ -861,23 +861,24 @@ export default class Item5e extends Item {
...
 
     // Commit pending data updates
-    const { actorUpdates, itemUpdates, resourceUpdates } = usage;
+    const { actorUpdates, itemUpdates, resourceUpdates, deleteIds } = usage;
     if ( !foundry.utils.isEmpty(itemUpdates) ) await item.update(itemUpdates);
-    if ( config.consumeQuantity && (item.system.quantity === 0) ) await item.delete();
+    if ( !foundry.utils.isEmpty(deleteIds) ) await this.actor.deleteEmbeddedDocuments("Item", [...deleteIds]);
     if ( !foundry.utils.isEmpty(actorUpdates) ) await this.actor.update(actorUpdates);
-    if ( resourceUpdates.length ) await this.actor.updateEmbeddedDocuments("Item", resourceUpdates);
+    if ( !foundry.utils.isEmpty(resourceUpdates) ) await this.actor.updateEmbeddedDocuments("Item", resourceUpdates);
```
[the remark this answers is not in the document]
Comment (krbz999) [reply module/documents/item.mjs:874]: It works for arrays now due to no longer being `isObjectEmpty` (since v10 I think?)... it just looked nicer, I admit.

File: module/data/item/templates/activated-effect.mjs (hunk 2 above)
Comment (krbz999) [reply module/data/item/templates/activated-effect.mjs:91]: I felt that we had agreed on it being useful but I'm happy to do that, lmk

File: module/documents/item.mjs (hunk 1 above)
Comment (krbz999) [reply module/documents/item.mjs]: This seems to work.

File: module/documents/item.mjs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Fyorl) [reply module/documents/item.mjs]: Yeah, that makes sense, since we accept the `"pact"` string as an option. I think just for ergonomics it's pretty easy for us to support `number` though as well, e.g.:
```js
Number.isInteger(config.slotLevel) ? config.slotLevel : config.slotLevel === "pact" ? as.spells.pact.level : parseInt(config.slotLevel.replace("spell", ""));
```
It's a bit better for backwards-compatibility too.

File: module/documents/item.mjs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Fyorl) [reply module/documents/item.mjs:884]: It's fine to use `isEmpty` then, just making sure I hadn't missed anything.

File: module/data/item/templates/activated-effect.mjs (hunk 2 above)
Comment (Fyorl) [reply module/data/item/templates/activated-effect.mjs]: Yeah, I think there's an issue for it already that we can add to milestone if it's not already in it. I'd just prefer they be split into separate PRs.

**Review by Fyorl — CHANGES_REQUESTED** [review]
I caught a couple more things on re-review, but otherwise this looks pretty much there. I will try for a faster turn around on any comments/changes here. Apologies for the delay.

File: module/documents/item.mjs [hunk 7]
```diff
@@ -784,25 +795,21 @@ export default class Item5e extends Item {
...
+    }
+    if ( Number.isNumeric(config.slotLevel) ) {
+      console.warn(`You are passing a numeric value to 'slotLevel' which expects a key from Actor5e#system#spells.`);
+      config.slotLevel = `spell${config.slotLevel}`;
+    }
+
+    config = foundry.utils.mergeObject(this._getUsageConfig(), config);
+
+    // Are any default values necessitating a prompt?
+    const needsConfiguration = Object.values(config).includes(true);
```
[the remark this answers is not in the document]
Comment (krbz999) [reply module/documents/item.mjs:812]: That is correct, yes. Since we will be using `null` for "the item is unable to do this", the follow-up 'prompt' work should set it to `false` (since it is able, but should not ask for it).

**Comment by Fyorl** [comment]
@krbz999 Any thoughts on [this](https://github.com/foundryvtt/dnd5e/pull/2400#discussion_r1358641259)? Do you have a strong objection to it?

File: module/documents/item.mjs (hunk 5 above)
[the remark this answers is not in the document]
Comment (krbz999) [reply module/documents/item.mjs]: > Yeah, that makes sense, since we accept the `"pact"` string as an option. I think just for ergonomics it's pretty easy for us to support `number` though as well, e.g.:
> 
> ```js
> Number.isInteger(config.slotLevel) ? config.slotLevel : config.slotLevel === "pact" ? as.spells.pact.level : parseInt(config.slotLevel.replace("spell", ""));
> ```
> 
> It's a bit better for backwards-compatibility too.

Since we're going for zero backwards compatibility already (`false` in the config will not be doing what it does now), I don't see much point in going half-way for this.

Unless I am mistaken, will we not at one point (unless we do so already) support custom spell progressions beyond `spellx` and `pact`? If so, seems like a string key is a one-size-fits-all.

[the remark this answers is not in the document]
Comment (Fyorl) [reply module/documents/item.mjs]: > Unless I am mistaken, will we not at one point (unless we do so already) support custom spell progressions beyond `spellx` and `pact`? If so, seems like a string key is a one-size-fits-all.

We transact spell slot levels throughout the codebase as numbers. It's not difficult to imagine our own code or 3rd-party code having the spell level as a number, but when passing it to this method they have to awkwardly cast it to a string with `level.toString()`. I think a one-size-fits-all solution should really allow for numbers.

[the remark this answers is not in the document]
Comment (krbz999) [reply module/documents/item.mjs]: Not married to either, so sure, not an issue.
